Refactor price_info

andrebeat opened the pull request:
This PR updates `price_info` to use `Fetch` instead of using hyper directly.
Additionally, the `Fetch` client is now cached to avoid creation on each
request.

I've tried to minimize the impact of these changes on the rest of the `ethcore`
module. The `price_info::Client` uses the `fetch::Client` directly instead of
being parameterized on `F: Fetch`, this is to avoid proliferating this type
parameter through the rest of `ethcore`. The `fetch::Client` is also created by
`price_info::Client` since it's not particularly easy to inject this dependency.
Currently this spawns 4 threads which will be mostly idle. I could add a
constructor to `fetch::Client` that allows defining the number of threads used
in the `CpuPool`.

Fixes #4205.

rphmeier commented:
I think @axelchalon was also working on this (you guys may want to comment on issues you are starting work on to avoid running into each other in the future)

andrebeat commented:
@rphmeier You're right, I'm sorry 😕 I should definitely have done that, especially considering that also happened on my last PR. @axelchalon let me know what's your progress on this and if necessary I'll close this PR.

rphmeier commented on ethcore/src/miner/price_info.rs:
```diff
@@ -15,94 +15,95 @@
-		});
+impl fmt::Debug for Client {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+		write!(fmt, "Client {{ api_endpoint: {} }}", self.api_endpoint)
```
maybe have the debug info indicate it's a price fetch client? We have another `Client` which is already very prolific throughout the codebase.

rphmeier commented on ethcore/src/miner/price_info.rs:
```diff
@@ -15,94 +15,95 @@
+impl Client {
+	pub fn new() -> Client {
+		let api_endpoint = "http://api.etherscan.io/api?module=stats&action=ethprice".to_owned();
+		let fetch = FetchClient::new().expect("Error starting fetch client");
```
it should return an error upon failure to start rather than aborting the whole program

andrebeat replied:
I agree, and the only reason I did this was to minimize the API changes in `ethcore`. Specifically, the `GasPricer` (which holds the `price_info`) is created from a `GasPricerConfig` using `Into` which assumes no failures.

tomusdrw replied:
Actually it would be best if the `FetchClient` was shared with other services.

debris commented:
Imo, if we are refactoring `price_info` it would be good to split it from `ethcore` to it's own crate.

axelchalon commented:
@andrebeat I hadn't started working on it, no worries! I'll make sure to comment on the issues more in the future!

tomusdrw commented on ethcore/src/miner/price_info.rs:
```diff
@@ -15,94 +15,95 @@
+	/// There was an error when trying to reach the API.
+	Fetch(String),
+	/// IO error when reading API response.
+	Io(String),
```
Why doesn't the errors just wrap `fetch::Error` and `io::Error`?

tomusdrw commented on ethcore/src/miner/price_info.rs:
```diff
@@ -15,94 +15,95 @@
-		});
+impl fmt::Debug for Client {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+		write!(fmt, "price_info::Client {{ api_endpoint: {} }}", self.api_endpoint)
```
`fmt.debug_struct` instead?

andrebeat commented:
@debris should this be moved to a root crate or should I move it to `util`? (or some other place).

andrebeat commented:
@tomusdrw @rphmeier I'll refactor this to inject the `FetchClient` dependency.

debris commented:
@andrebeat root directory is ok

andrebeat commented:
I think I have addressed all your comments. I also changed `price_info` to use `serde_json` instead of `rustc_serialize`.

tomusdrw commented on parity/params.rs:
```diff
@@ -226,15 +227,17 @@ impl Default for GasPricerConfig {
+						usd_per_tx: usd_per_tx,
+						recalibration_period: recalibration_period,
+					},
+					fetch)
```
nitpick: imho it looks bettter like this:
```rust
  ...
  },
  fetch,
)
...
```

tomusdrw commented on ethcore/Cargo.toml:
```diff
@@ -32,6 +32,7 @@ ethjson = { path = "../json" }
 ethkey = { path = "../ethkey" }
 ethstore = { path = "../ethstore" }
 evmjit = { path = "../evmjit", optional = true }
+fetch = { path = "../util/fetch" }
```
To reduce the number of dependencies IMHO we could re-export fetch from `price-info`.

tomusdrw commented on Cargo.toml:
```diff
@@ -42,6 +42,7 @@ ethcore-light = { path = "ethcore/light" }
 ethcore-logger = { path = "logger" }
 ethcore-stratum = { path = "stratum" }
 ethkey = { path = "ethkey" }
+fetch = { path = "util/fetch" }
```
Main crate is already able to instantiate `FetchClient` using a re-export from `hash-fetch` crate.

tomusdrw commented on price-info/src/lib.rs:
```diff
@@ -0,0 +1,33 @@
+
+#![warn(missing_docs)]
+#![cfg_attr(feature="benches", feature(test))]
+#![cfg_attr(feature="dev", feature(plugin))]
```
No features are not listed in the `Cargo.toml` so this seems to be redundant.

tomusdrw commented on price-info/src/price_info.rs:
```diff
@@ -0,0 +1,145 @@
+	use price_info::{Client, PriceInfo};
+
+	#[test] #[ignore]
+	fn should_get_price_info() {
```
Actually if `Client` would be generic over `Fetch` you would be able to write proper tests for this service.

andrebeat replied:
Sure, I'll update it 👍

andrebeat commented:
@tomusdrw I've fixed the re-exports. I also made the `Client` generic over `Fetch` and added some tests. I added a default value to the `Fetch` type parameter to avoid "polluting" the `ethcore` API.

rphmeier commented on util/fetch/src/client.rs:
```diff
@@ -61,6 +61,14 @@ pub trait Fetch: Clone + Send + Sync + 'static {
+	/// Spawn the future in context of this `Fetch` thread pool as "fire and forget", i.e. dropping this future without
+	/// canceling the underlying future.
+	/// Implementation is optional.
+	fn forget<F, I, E>(&self, _: F) where
```
why is implementation optional?

gavofyork replied:
@andrebeat ?

andrebeat replied:
@rphmeier You're right, it should not be optional. I think the same may be true for the `process` method.
